### ISSUE TITLE
SRV for multi-port applications

### ISSUE THREAD

**Issue by alberts** [mesosphere/mesos-dns#61]
Hey, Went to the meetup in SF tonight. Thanks for the presentation. IIRC (I could be wrong), in @guenter's demo he looked up a SRV of the form _foo.bar._tcp.mesos and it successfully resolved, but all the docs in this repo seem to contain examples that conform to RFC 2782, so maybe that was some kind of typo or old version of Mesos DNS.

Anyway, I was wondering if you've given any thought to how multi-port applications will be exposed via SRV? It would seem that a naming convention where the port becomes the service part, e.g., _port[012]._tcp.foo.bar.mesos could work.

Also, have you given any thought to how you might populate priorities? It might be nice to be able to tag tasks running in DC1 with priority 1 and tasks running in DC2 with priority 10, so that if all the DC1 nodes/tasks are unreachable (even if they are still in DNS), applications can know to switch to DC2. This is probably a bit of an edge case but it would be interesting to hear your thoughts.

We're keen to deploy Mesos DNS ASAP, hence the questions. Thanks for open sourcing this.

**Comment by kozyraki** [comment]
Thanks for the feedback Albert. Let's separate issues here:

***Apps with multiple ports***:  If you have an app that listens to a number of ports (e.g., 8181 and 8282), we can create multiple SRV records. So, if you lookup the service name (e.g., _app._tcp.domain) you will get multiple answers, same IP for both, but different ports (8181 and 8282). I assume this is useful for people who know the service name but don't know the actual ports used (may not be the default ones). Let me know if this is useful for you. If someone knows the ports and wants this as part of the service name, then it is a different discussion (see below). 

***Naming scheme***: The current naming scheme in mesos-dns is simple and limited. My hope is that the 0.22.0 release of mesos will include the updates I made for Jira MESOS-2150. This will allow frameworks to pass much more interesting information to Mesos about how they want their task to be discovered (names, env/loc/ variables, general labels). Once we have this, we can support much more interesting names for both A and SRV records. My suspicion is that there is no single way to do this and we need to support some way for users to select. Assuming MESOS-2150, what kind of naming scheme(s) would work best for you? Give us some guidance.

**Comment by mindscratch** [comment]
+1 for "Apps with multiple ports"

**Comment by alberts** [comment]
Re: Apps with multiple ports: don't know if multiple SRV records with different ports for the same service name makes sense. If I get back two ports for the same host, how do I know which one is the "first" port for my application, and which is the second? Are ports assigned by Mesos guaranteed to be in ascending order from port0...portN? Seems to me like a "service" with two ports could be seen as two services, _port0._tcp.service.domain and _port1._tcp.service.domain.

**Comment by kozyraki** [comment]
@alberts This issue goes a beyond Mesos-DNS. By the time Mesos-DNS gets the information from the master (imagine MESOS-2150 is there), port assignment has happened. All Mesos-DNS is create various types of names (like _port0._tcp.service.domain). You are asking how does port assignment happen and who is responsible for updating the ports fields on the service discovery info introduced by Mesos-2150. Some frameworks handle port assignment (e.g., see discussion at https://github.com/mesosphere/marathon/pull/384 for Marathon). 

So, let's break the problem into two:
a) does a framework like Marathon allow you to specify the port requirements of your app and the names they should get? 
b) does Mesos-DNS name correctly the services afterwards?

**Comment by mindscratch** [comment]
I can respond to "a" based on my experience: marathon lets you specify the ports but they're each associated with a single name. I believe the ports given to marathon are kept in order, so doing a request and getting back multiple ports could be ok, as I could use a convention like "last port" is the "http" port or last port is the "service port", depends on what I need for the particular service.

**Comment by kozyraki** [comment]
I filed this with Marathon. Add to it please.
https://github.com/mesosphere/marathon/issues/1112

**Comment by tsenart** [comment]
Is this solved by frameworks providing discovery info?

**Comment by jdef** [comment]
TL;DR: no

It isn't solved yet in mesos-dns. part of the solution is for frameworks to *name* ports in their DiscoveryInfo. Another part of the solution is for mesos-dns to use that name in SRV lookups somehow. There have been several proposals discussed and there hasn't been 100% consensus across the team. For example, a recent proposal attempted to standardize the SRV namespace:
```
_{task}._{proto}.{framework}.{domain}
```

in favor of the following (consistent `_service` namespace: `_ep`, `_nm` replaces `_{task}`):
```
SRV _ep._{proto}.{task}.{framework}.{domain} // ep: stands for endpoint, get you all task ip:port's 
SRV _nm._{proto}.{di-port-name}.{task}.{framework}.{domain} // nm: stands for "named DI port"
```

It's hard to imagine actually dropping the former since people are pretty dependent upon it now. However that doesn't stop us from implementing the `_nm` SRV records for named DI ports on top of what we already have.

Another proposal hinted at a [DNS-SD](https://tools.ietf.org/html/rfc6763#section-13.2) like representation that used PTR records (which catalog SRVs):
```
PTR _{di-port-name}._sub._{task}._{proto}.{framework}.{domain} --> _nm._{proto}.{di-port-name}.{task}.{framework}.{domain}
```

and it's also not hard to imagine addressing a specific task port via DNS-SD SRV *(probably the most natural evolution of what we have now)*:
```
SRV {di-port-name}._{task}._{proto}.{framework}.{domain}

example for k8s/dcos service components:
apiserver-healthz._kubernetes._tcp.marathon.mesos
apiserver-api._kubernetes._tcp.marathon.mesos
scheduler-healthz._kubernetes._tcp.marathon.mesos
scheduler-api._kubernetes._tcp.marathon.mesos
...
```

I don't think we have rules (yet) that take into account DI visibility constraints either.

**Comment by doctapp** [comment]
+1

**Comment by bryanstephens** [comment]
+1

**Comment by quofelix** [comment]
Another use case for our system (Quobyte). Quobyte already finds its registry nodes via SRV lookups. However, as we export RPC and HTTP ports, there are SRV records that contain both port numbers and the resulting SRV records are not usable. So +1 for anything that lets me specify the port by index or even name.

**Comment by metskem** [comment]
+1

**Comment by abraithwaite** [comment]
+1

**Comment by user31459** [comment]
+1

**Comment by sargun** [comment]
@aquamatthias When is Marathon going to set DiscoveryInfos on ports?

**Comment by aquamatthias** [comment]
Planned for Marathon 0.16 (current release)

**Comment by eLvErDe** [comment]
+1

We're having the issue with RabbitMQ instances, no way to identify app port from management interface. Right now my error-proof connection class handle it anyway but it's really dirty :)

Thanks in advance!

**Comment by dry4ng** [comment]
+1

**Comment by jdef** [comment]
mesos-dns generates SRVs per port when DiscoveryInfo is present, closing this out.
